### PR TITLE
CMP: add `-profile` and extend `-geninfo` option

### PR DESCRIPTION
**Pull request by DDvO**
* CMP lib and app: add optional `certProfile` request message header and respective `-profile` option, 
   carved out from #18240
* CMP app: make `-geninfo` option accept multiple ITAVs and support string values besides integers
* Add missing getter functions `OSSL_CMP_{CTX,HDR}_get0_geninfo_ITAVs()` to CMP API

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by DDvO** [comment]
CI failure on **no-thread-pool** (`80-test_cms.t`) is unrelated.

**Comment by DDvO** [comment]
@openssl/committers please also provide a review of this PR soon,
such that it will make it into 3.2.
Part of the code, namely the certProfile aspect, has already been reviewed as part of #18240 mentioned above.

**Comment by DDvO** [comment]
Rebased to fix trivial merge conflict in `libcrypto.num`.

Can we still get this feature into 3.2?

**Comment by DDvO** [comment]
Rebased to fix merge conflicts.

**Comment by DDvO** [comment]
@openssl/committers can we please also get this handled soon, for inclusion in 3.2?

This PR does not contain difficult stuff and contains many positive and negative test cases demonstrating that the new features work as expected.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Ping @openssl/committers for 2nd review

**Comment by DDvO** [comment]
Asking again @openssl/committers for 2nd review/approval.
Should be rather straightforward to do.

**Comment by DDvO** [comment]
Ping @openssl/committers for 2nd review

File: doc/man3/OSSL_CMP_ITAV_set0.pod
Comment (tmshort) [inline doc/man3/OSSL_CMP_ITAV_set0.pod:54]:
```suggestion
On success, ownership of the list is with the new B<OSSL_CMP_ITAV> structure.
```

File: include/openssl/cmp.h.in
Comment (tmshort) [inline include/openssl/cmp.h.in:265]: Put on a single line?
Or, indent the second line?

File: crypto/cmp/cmp_hdr.c
Comment (tmshort) [inline crypto/cmp/cmp_hdr.c:78]: Inconsistency with other functions that require an argument (e.g. most `SSL_` functions).

File: apps/cmp.c
Comment (tmshort) [inline apps/cmp.c:1862]: Should this `OSSL_assert`, instead of ignoring the error?

File: doc/man3/OSSL_CMP_ITAV_set0.pod
Comment (DDvO) [reply doc/man3/OSSL_CMP_ITAV_set0.pod]: Oops, fixed.

File: include/openssl/cmp.h.in
Comment (DDvO) [reply include/openssl/cmp.h.in]: Both `check-formal.pl` and the Emacs C mode are fine with the above indentation,
and adding an indentation in the 2nd line would not be correct according to both tools.
Anyway, changed to
```
OSSL_CMP_ITAV *OSSL_CMP_ITAV_new0_certProfile(STACK_OF(ASN1_UTF8STRING)
                                              *certProfile);
```
which may look better and is also correct.

File: crypto/cmp/cmp_hdr.c
Comment (DDvO) [reply crypto/cmp/cmp_hdr.c:78]: I guess you mean the NULL check for the parameter.
This style is consistent with all the CMP API functions.
When I contributed them, I was told that all parameters of API functions shall be checked and to add an error queue entry on failure.

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c:1862]: There is no `OSSL_assert`, so I presume you mean `ossl_assert`.
This is only available internally,
while in `apps/` there are many such uses of `(void)` where nothing can go wrong.
In this case there can be no OOM error due to `sk_ASN1_UTF8STRING_new_reserve(NULL, 1)`.

**Comment by DDvO** [comment]
Thank you @tmshort for your review comments. I've handled all of them.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
@tmshort can you approve now?

**Comment by DDvO** [comment]
@tmshort ok now?

**Comment by DDvO** [comment]
@tmshort did you see that your review comments have been handled?

**Comment by DDvO** [comment]
@openssl/committers since @tmshort seems unavailable since two weeks already,
can anyone else please provide the missing 2nd approval?

**Comment by tom-cosgrove-arm** [comment]
@openssl/committers I don't have the bandwidth to review this in the next week with other commitments - is any other committer available for review?

**Comment by DDvO** [comment]
@tmshort looks like meanwhile your are available again?
Can you give your approval now?

**Comment by DDvO** [comment]
@openssl/committers who can help with the 2nd approval?

**Comment by tom-cosgrove-arm** [comment]
I do not have bandwidth to look at this one, sorry

**Comment by DDvO** [comment]
Added a little grammar fix in `openssl-cmp.pod.in`,
which came up in a backport PR: https://github.com/openssl/openssl/pull/21138#discussion_r1326900978.

**Comment by DDvO** [comment]
Asking again for the missing 2nd approval - 
hopefully now someone of the @openssl/committers has some bandwidth for it?

**Comment by DDvO** [comment]
Rebased to fix merge conflicts in `util/libcryto.num`.

Added a fixup to bump the new OpenSSL version references and add a `CHANGES.md` entry.

**Comment by DDvO** [comment]
Since OpenSSL 3.2 is meanwhile out, I had to update this PR a little.

@openssl/committers **_could you please give the final blessing to this PR soon._**
It already received a review by @tmshort and an approval by @t8m in August,
so the 2nd review should not take too much effort.

**Review by t8m — APPROVED** [review]

File: doc/man3/OSSL_CMP_ITAV_set0.pod
Comment (slontis) [inline doc/man3/OSSL_CMP_ITAV_set0.pod:60]: Is there somewhere in the docs where you define what the types actually are?
When you say certProfile, this seems like  an internal detail (i.e. this is an opaque structure). Do you list the OID's somewhere?

File: doc/man1/openssl-cmp.pod.in
Comment (slontis) [inline doc/man1/openssl-cmp.pod.in:1404]: as was?

File: crypto/cmp/cmp_hdr.c
Comment (slontis) [reply crypto/cmp/cmp_hdr.c:78]: I think he means the function signature i.e.
int func(*obj, **ret)

Comment (slontis) [reply crypto/cmp/cmp_hdr.c:78]: Probably a bit late now :)

File: apps/lib/cmp_mock_srv.c
Comment (slontis) [inline apps/lib/cmp_mock_srv.c:250]: Is this bad if itavs is NULL? i.e. sk_*_num returns -1.

File: apps/lib/cmp_mock_srv.c
Comment (slontis) [inline apps/lib/cmp_mock_srv.c:264]: Could this be null?

File: apps/cmp.c
Comment (slontis) [inline apps/cmp.c:1862]: What do you mean it must succeed, _push can do a malloc?

File: apps/cmp.c
Comment (slontis) [inline apps/cmp.c:1950]: I assume you dont care that this could build a partial list if it fails after the first loop?

File: doc/man3/OSSL_CMP_ITAV_set0.pod
Comment (DDvO) [reply doc/man3/OSSL_CMP_ITAV_set0.pod:60]: For most part, except for applicability of `OSSL_CMP_ITAV_get0_certProfile()`,  
users of the two new functions can consider the type components of `OSSL_CMP_ITAV` (where ITAV is short for InfoTypeAndValue) as internal detail.

These types are represented by OIDs, and those needed for CMPv2+ have been introduced by #15778 in `include//openssl/obj_mac.h`.
The specific `id-it-certProfile` has been defined by https://datatracker.ietf.org/doc/html/rfc9480#name-new-section-5113-certprofil as `OBJECT IDENTIFIER ::= {id-it 21}`.

I've added references in the `NOTES` section as follows:
```
OIDs to use as types in B<OSSL_CMP_ITAV> can be found at
L<https://datatracker.ietf.org/doc/html/rfc9480#section-4.2.2>.
The respective OpenSSL NIDs, such as B<NID_id_it_certProfile>,
are defined in the F<< <openssl/obj_mac.h> >> header file.
```

File: doc/man1/openssl-cmp.pod.in
Comment (DDvO) [reply doc/man1/openssl-cmp.pod.in]: Oops, removed the 'as'.

File: apps/lib/cmp_mock_srv.c
Comment (DDvO) [reply apps/lib/cmp_mock_srv.c:250]: The case `itavs == NULL` is not a problem - 
both conceptually and for the given `for(...)` loop, having no ITAVs is equivalent to having a empty list of them.

File: apps/lib/cmp_mock_srv.c
Comment (DDvO) [reply apps/lib/cmp_mock_srv.c]: Good question - I'm not sure (anymore) and `ASN1_UTF8STRING` is not documented.
So to be on the safe side, I've added NULL checks.

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c]: Due to the `sk_ASN1_UTF8STRING_new_reserve(NULL, 1)`, there will be no malloc call and thus it cannot fail.
I've improved the comment to
```
   /* Due to sk_ASN1_UTF8STRING_new_reserve(NULL, 1), this surely succeeds: */
```

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c:1951]: Right. 
If this fails (likely due to malloc failure) and elements were successfully added to the list in `ctx` in earlier loop iterations,
the overall function and thus the app fails, whereby the `ctx` gets deallocated, freeing any ITAVS pushed before.

**Comment by DDvO** [comment]
Thanks a lot @slontis for reviewing and your good. comments!
I've handled all of them.

**Review by slontis — APPROVED** [review]
@t8m can you reapprove if you are happy with the changes.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Merged - thank you @t8m, @tmshort, and @slonts for your reviews and approvals.

**Pull request closed without merging**
